xdsl/dialects/snitch_runtime.py

```diff
@@ -398,14 +403,185 @@ def __init__(self, transfer_id: Operation | SSAValue):
 
 
 @irdl_op_definition
-class DmaWaitAllOp(SnitchRuntimeBarrier):
+class DmaWaitAllOp(NoOperandNoResultBaseOp):
     """
     Block until all operations on the DMA cease
     """
 
     name = "snrt.dma_wait_all"
 
 
+"""
+The number of data movers determines the number of
+independent memory address patterns a core can keep track
+of. Since the data movers are tied to individual registers,
+there need to be at least the same number of registers with
+stream semantics as there are data movers. Multiple SSRs
+may address the same data mover, for example to use the
+data mover both in integer and FP instructions.
+for more info check https://arxiv.org/pdf/1911.08356.pdf
+
+The different SSR data movers.
+    SNRT_SSR_DM0 = 0,
+    SNRT_SSR_DM1 = 1,
+    SNRT_SSR_DM2 = 2,
+"""
+ssr_dm = i32
+
+"""
+The different dimensions - those determine how many levels of nesting a loop can have.
+The snitch system handles cases with up to 4, but this can be extended.
+    SNRT_SSR_1D = 0,
+    SNRT_SSR_2D = 1,
+    SNRT_SSR_3D = 2,
+    SNRT_SSR_4D = 3,
+"""
+ssr_dim = i32
+
+
+class SsrLoopBaseOp(SnitchRuntimeBaseOp, ABC):
+    """
+    Configure an SSR data mover for an n-dimensional loop nest.
+    """
+
+    name = "snrt.ssr_loop_1d"
+    data_mover: Operand = operand_def(ssr_dm)
+    bounds: VarOperand = var_operand_def(IndexType)
+    indices: VarOperand = var_operand_def(IndexType)
+
+    irdl_options = [AttrSizedOperandSegments()]
+
+    def verify_(self) -> None:
+        if len(self.bounds) != len(self.indices):
+            raise VerifyException(
+                f"the length of bounds ({len(self.bounds)}) and indices ({len(self.indices)}) must be equal."
+            )
+
+    def __init__(
+        self,
+        data_mover: Operation | SSAValue,
+        bounds: Sequence[Operation | SSAValue],
+        indices: Sequence[Operation | SSAValue],
+    ):
+        super().__init__(operands=[data_mover, bounds, indices], result_types=[])
+
+
+@irdl_op_definition
+class SsrLoop1dOp(SsrLoopBaseOp):
+    """
+    Configure an SSR data mover for a 1D loop nest.
+    """
+
+    name = "snrt.ssr_loop_1d"
```

Comment:
I don't think this makes sense, it doesn't actually restict the number of operations. Do we need the 1/2/3/4d operations to exist explicitly?

Reply:
It feels like there should be a hook to create the ssr_loop_xd func call name dynamicaly based on the number of operands. There should probably be a validator that checks that there are 1-4 of them.

Reply:
@superlopuh I agree. right now people can still make a ssr_loop_1d op with 2 dimensions if they like.
That should not be possible.
It makes sense to have 1 - 4D explicitly registered right now though, since these are the ones provided by the C implementation of the snitch runtime, so it makes sense to reflect them in this dialect for now.

Reply:
What's the point of the subclass operations with the names? I don't see how they are helpful.

Reply:
hmm, so what should I do?

Reply:
I think I'd go with @JosseVanDelm's commment for now; let's aim this close to the C rt and we can update later.

Reply:
I'm happy as soon as there's a check that guarantees that there's no way to construct an `ssr_loop_1d` op with 2 dimensions, and a unit tests that checks for this.